### PR TITLE
feat: adding cross_contract_calls integration test for the V1->V2 resharding

### PR DESCRIPTION
**Pull request by wacban**
The goal of this PR is to refactor the existing cross_contract_calls integration test so that it can be reused for testing the new resharding and do that. I have encountered numerous issues when trying to adapt that test because enabling the new resharding also enables all of other protocol features since simple nightshade. The test wasn't quite ready for it. I managed to fix a few issues but not all. I'm leaving it unfinished for now as I suspect the current issue may be due to lack of flat storage support so it's better to wait for that. I marked the new test as #[ignore]. I would still like to merge in all of the refactoring, fixes and improvements I made so far and I will pick it up again once flat storage is supported in resharding. 

Some of the new features are:
- I refactored the cross_contract_calls so that it can handle either resharding type.
- I allowed not all clients to process all transactions but also route them. This was needed only before I set all clients to track all shards but is correct, handles both cases and it more future proof. 
 
Some of the challenges were:
- The test was failing with missing chunk error. This was because the "Partial Encoded Chunk Forward" messages were not correctly handled in the test infra. I fixed this issue in TestEnv::process_partial_encoded_chunks. There are two parts of the fix:
  - handle the forwards
  - keeping looping until all messages are processed - this is needed because one message can trigger sending of another message
- The protocol version was not getting upgraded and because of that the resharding was not getting triggered (or at least not deterministically where I wanted it to happen). This was because of unlucky block producer assignment where one block producer (out of 4) does not actually get to produce any blocks in the (rather short) first epoch. When that happens voting for the new protocol version is stuck at 3/4 = 75% votes which is lower than the default threshold. I lowered the threshold in the test to work around that. 
- The state sync process was getting triggered and failing because of lack of some event loop or other async framework setup. I fixed that by setting all nodes to track all shards. (not an issue in V0->V1 because when there is 1 shard everyone tracks it and apparently the node keep tracking the new shards as well).

### CODE REVIEW

**Comment by robin-near** [comment]
> keeping looping until all messages are processed - this is needed because one message can trigger sending of another message

This is what the TestLoop does. It requires of course some sort of a migration to make this test work in that framework.

File: chain/chunks/src/lib.rs
Comment (wacban) [inline chain/chunks/src/lib.rs:609]: this is better reviewed with whitespace hidden

File: chain/chunks/src/logic.rs
Comment (wacban) [inline chain/chunks/src/logic.rs:156]: this is removed by accident I will revert it

File: chain/client/src/test_utils.rs
Comment (wacban) [inline chain/client/src/test_utils.rs:1892]: @robin-near answering your comment from the main discussion here

In the now inner loop processing of a PEC may lead to PECF send to clients that were already processed and those messages would get lost. This is why I added the outer loop.

File: chain/chain/src/chain.rs
Comment (shreyan-gupta) [inline chain/chain/src/chain.rs:2764]: I'm wondering whether it's ever possible for us to hit this? I checked the implementation and this is what we have
```
    pub fn will_shard_layout_change(&self, parent_hash: &CryptoHash) -> Result<bool, EpochError> {
        let epoch_id = self.get_epoch_id_from_prev_block(parent_hash)?;
        let next_epoch_id = self.get_next_epoch_id_from_prev_block(parent_hash)?;
        let shard_layout = self.get_shard_layout(&epoch_id)?;
        let next_shard_layout = self.get_shard_layout(&next_epoch_id)?;
        Ok(shard_layout != next_shard_layout)
    }
 ```
 Wondering which one of these functions can fail?

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (shreyan-gupta) [inline integration-tests/src/tests/client/sharding_upgrade.rs:660]: Is `.track_all_shards()` a fix for this issue?

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (shreyan-gupta) [inline integration-tests/src/tests/client/sharding_upgrade.rs:206]: nit: can directly call `.entered()` here? Can add some text to the span as well or remove altogether.

File: chain/client/src/test_utils.rs
Comment (shreyan-gupta) [reply chain/client/src/test_utils.rs:1892]: Didn't completely understand what's going on here. Ignoring for now.

**Comment by wacban** [comment]
> > keeping looping until all messages are processed - this is needed because one message can trigger sending of another message
> 
> This is what the TestLoop does. It requires of course some sort of a migration to make this test work in that framework.

Ah, I see now what you mean, there is an existing framework for doing this in tests. I will have a look but for now I would rather merge this as is and follow up with improvements later.

**Review by shreyan-gupta — APPROVED** [review]

File: chain/chain/src/chain.rs
Comment (wacban) [reply chain/chain/src/chain.rs:2764]: It is rather unlikely, I think the rocksdb would have to fail itself. Still worth handling somehow if we can. It's not a priority right now as we're focused on the happy path but adding a TODO so that we can come back to it.

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (wacban) [reply integration-tests/src/tests/client/sharding_upgrade.rs:660]: yes

File: chain/client/src/test_utils.rs
Comment (wacban) [reply chain/client/src/test_utils.rs:1892]: old implementation only handled the PartialEncodedChunkMessage (PEC)
new implementation also handles PartialEncodedChunkForward (PECF)
new implementation keeps looping as long as there are any messages processed at all in the previous iteration - because receiving a PEC triggers sending PECF and we want to make sure we handle all of the messages

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (walnut-the-cat) [inline integration-tests/src/tests/client/sharding_upgrade.rs:230]: simply moved to Self::process_tx(env, tx);

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (walnut-the-cat) [inline integration-tests/src/tests/client/sharding_upgrade.rs:211]: this is pretty interesting way to assign value :o

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (walnut-the-cat) [inline integration-tests/src/tests/client/sharding_upgrade.rs:244]: just moved into {} block for block var

**Review by walnut-the-cat — COMMENTED** [review]
nit: looking at the changes, PR includes a lot of refactoring which is not necessarily directly related to title of the PR. It may be better to separate refactoring part out as its own PR for better review-ability

**Comment by wacban** [comment]
> nit: looking at the changes, PR includes a lot of refactoring which is not necessarily directly related to title of the PR. It may be better to separate refactoring part out as its own PR for better review-ability

Generally I agree but I struggle to make that work in github. I couldn't get PRs to be stacked on top of each other, if anyone knows how to do it right I'm all ears. Some of the refactoring is relevant here, like process_tx is replaced some other code in second location.

File: chain/chunks/src/lib.rs
Comment (wacban) [reply chain/chunks/src/lib.rs:609]: No, that's just it, it reduces the nesting of the following code and makes some lines fit into a single line - both make it more readable.

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (wacban) [reply integration-tests/src/tests/client/sharding_upgrade.rs:211]: yep, I just wanted the span to be contained in a dedicated scope, probably might have move it to a method but didn't think of it at the time

**Comment by jakmeier** [comment]
> > nit: looking at the changes, PR includes a lot of refactoring which is not necessarily directly related to title of the PR. It may be better to separate refactoring part out as its own PR for better review-ability
> 
> Generally I agree but I struggle to make that work in github. I couldn't get PRs to be stacked on top of each other, if anyone knows how to do it right I'm all ears. Some of the refactoring is relevant here, like process_tx is replaced some other code in second location.

I usually just submit the refactoring as a separate PR, with a comment that it's prep for another PR. Usually I don't submit the second PR yet because it requires the changes from the first PR.

When I want to submit them both together for better context, I like to submit a dummy PR from the second branch targeting the first PR branch. This way people can review the changes of the second PR in isolation.

In any case, changes after reviewing PR0 will mess things up for PR1 and after the squash-merge you will have to rebase + force-push PR1 anyway. So I usually just wait with PR1 until PR0 has landed.

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (jakmeier) [inline integration-tests/src/tests/client/sharding_upgrade.rs:1002]: Is there an issue you could link here where readers can figure out or ask about the progress on fixing this?

**Review by jakmeier — APPROVED** [review]
Mostly looks ok :)

Tbh, some of the refactored code could probably use more refactoring. But there could always be more refactoring. Your changes look like an overall improvement to me.

File: chain/epoch-manager/Cargo.toml
Comment (wacban) [reply chain/epoch-manager/Cargo.toml:21]: Yeah I usually use collect_vec when printing debug logs. How would you feel about leaving that in?

Comment (jakmeier) [reply chain/epoch-manager/Cargo.toml:21]: Generally speaking, I don't really like adding dependencies that aren't used in committed code. It's the kind of thing I personally might create a PR, just to remove a dead dependency. So at least a comment why it's in would be useful to prevent that sort of engineers canceling out each others "improvements".

But that said, itertools wouldn't be the worst to have pulled in for no reason, so I won't stop you from adding it if you hate writing `collect::<Vec<_>>` that much :D

Comment (wacban) [reply chain/epoch-manager/Cargo.toml:21]: Ok, thanks, I'll leave that in and add a comment.

File: integration-tests/src/tests/client/sharding_upgrade.rs
Comment (wacban) [reply integration-tests/src/tests/client/sharding_upgrade.rs:1011]: You never know, maybe versino is Spanish for version? ;)
Thanks, good catch, will fix soon.

**Pull request merged**
